src/server/socksServer.ts

```diff
@@ -0,0 +1,304 @@
+                if (b % 2 === 0 && b > 0)
+                  ipv6str += ':';
+                ipv6str += (addr[b] < 16 ? '0' : '') + addr[b].toString(16);
+              }
+              this._info.dstAddr = ipv6str;
+            } else {
+              this._info.dstAddr = this._dstAddr.toString();
+            }
+            this._info.dstPort = this._dstPort;
+            this._phase++;
+            this._readyResolve();
+            return;
+          }
+          i++;
+          break;
+        default:
+          assert(false);
+      }
+    }
+  }
+
+  private _authWithoutPassword(socket: net.Socket) {
+    socket.write(Buffer.from([0x05, 0x00]));
+  }
+
+  async ready(): Promise<{ info: SocksConnectionInfo, forward: () => void, intercept: () => SocksInterceptedHandler }> {
+    await this._ready;
+    return {
+      info: this._info,
+      forward: () => {
+        const dstSocket = new net.Socket();
+        this._socket.on('close', () => dstSocket.end());
+        this._socket.on('end', () => dstSocket.end());
+        dstSocket.setKeepAlive(false);
+        dstSocket.on('error', (err: NodeJS.ErrnoException) => writeSocksSocketError(this._socket, String(err)));
+        dstSocket.on('connect', () => {
+          this._socket.write(BUF_REP_INTR_SUCCESS);
+          this._socket.pipe(dstSocket).pipe(this._socket);
+          this._socket.resume();
+        }).connect(this._info.dstPort, this._info.dstAddr);
+      },
+      intercept: (): SocksInterceptedHandler => {
+        return new SocksInterceptedHandler(this._socket);
+      },
+    };
+  }
+}
+
+export class SocksInterceptedHandler {
+  socket: net.Socket;
+  constructor(socket: net.Socket) {
+    this.socket = socket;
+  }
+  connected() {
+    this.socket.write(BUF_REP_INTR_SUCCESS);
+    this.socket.resume();
+  }
+  error(error: string) {
+    this.socket.resume();
+    writeSocksSocketError(this.socket, error);
+  }
+  write(data: Buffer) {
+    this.socket.write(data);
+  }
+  end() {
+    this.socket.end();
+  }
+}
+
+function writeSocksSocketError(socket: net.Socket, error: string) {
+  if (!socket.writable)
+    return;
+  socket.write(BUF_REP_INTR_SUCCESS);
+
+  const body = `Could not connect: ${error}`;
+  socket.end([
+    'HTTP/1.1 502 OK',
+    'Connection: close',
+    'Content-Type: text/plain',
+    'Content-Length: ' + Buffer.byteLength(body),
+    '',
+    body
+  ].join('\r\n'));
+}
+
+type IncomingProxyRequestHandler = (info: SocksConnectionInfo, forward: () => void, intercept: () => SocksInterceptedHandler) => void
+
+export class SocksProxyServer {
+  public server: net.Server;
+  constructor(incomingMessageHandler: IncomingProxyRequestHandler) {
+    this.server = net.createServer(this._handleConnection.bind(this, incomingMessageHandler));
+  }
+
+  public listen(port: number, host?: string) {
+    this.server.listen(port, host);
+  }
+
+  _handleConnection = async (incomingMessageHandler: IncomingProxyRequestHandler, socket: net.Socket) => {
```

Comment:
Make this a regular method, not an arrow function please. Same for all other methods.